internal/controllers/machinedeployment/machinedeployment_status.go

```diff
@@ -0,0 +1,502 @@
+
+	setMachinesReadyCondition(ctx, s.machineDeployment, machines)
+	setMachinesUpToDateCondition(ctx, s.machineDeployment, machines)
+
+	setRemediatingCondition(ctx, s.machineDeployment, machinesToBeRemediated, unHealthyMachines)
+
+	setDeletingCondition(ctx, s.machineDeployment, s.machineSets, machines, s.getAndAdoptMachineSetsForDeploymentSucceeded)
+
+	return retErr
+}
+
+// setReplicas sets replicas in the v1beta2 status.
+// Note: this controller computes replicas several time during a reconcile, because those counters are
+// used by low level operations to take decisions, but also those decisions might impact the very same the counters
+// e.g. scale up MachinesSet is based on counters and it can change the value on MachineSet's replica number;
+// as a consequence it is required to compute the counters again before calling scale down machine sets,
+// and again to before computing the overall availability of the Machine deployment.
+func setReplicas(machineDeployment *clusterv1.MachineDeployment, machineSets []*clusterv1.MachineSet) {
+	if machineDeployment.Status.V1Beta2 == nil {
+		machineDeployment.Status.V1Beta2 = &clusterv1.MachineDeploymentV1Beta2Status{}
+	}
+
+	machineDeployment.Status.V1Beta2.ReadyReplicas = mdutil.GetV1Beta2ReadyReplicaCountForMachineSets(machineSets)
+	machineDeployment.Status.V1Beta2.AvailableReplicas = mdutil.GetV1Beta2AvailableReplicaCountForMachineSets(machineSets)
+	machineDeployment.Status.V1Beta2.UpToDateReplicas = mdutil.GetV1Beta2UptoDateReplicaCountForMachineSets(machineSets)
+}
+
+func setAvailableCondition(_ context.Context, machineDeployment *clusterv1.MachineDeployment, getAndAdoptMachineSetsForDeploymentSucceeded bool) {
+	// If we got unexpected errors in listing the machine sets (this should never happen), surface them.
+	if !getAndAdoptMachineSetsForDeploymentSucceeded {
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:    clusterv1.MachineDeploymentAvailableV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineDeploymentAvailableInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	// Surface if .spec.replicas is not yet set (this should never happen).
+	if machineDeployment.Spec.Replicas == nil {
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:   clusterv1.MachineDeploymentAvailableV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineDeploymentAvailableWaitingForReplicasSetV1Beta2Reason,
+		})
+		return
+	}
+
+	// Surface if .status.v1beta2.availableReplicas is not yet set.
+	if machineDeployment.Status.V1Beta2 == nil || machineDeployment.Status.V1Beta2.AvailableReplicas == nil {
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:   clusterv1.MachineDeploymentAvailableV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineDeploymentAvailableWaitingForAvailableReplicasSetV1Beta2Reason,
```

Comment:
If I see correctly this is unreachable code.

* If getAndAdoptMachineSetsForDeploymentSucceeded is false => we never get here (l.91)
* if getAndAdoptMachineSetsForDeploymentSucceeded is true => replicas will be set

Considering this and if you want to keep the check, can we just make this an InternalError instead of surfacing a separate reason?

Reply:
> getAndAdoptMachineSetsForDeploymentSucceeded is true => replicas will be set

We can't say this for sure. e.g. the MS has not been reconciled yet and so MS.Status.AvailableReplicas is nil, when we compute replicas sum of nil > nil and this MD.Status.AvailableReplicas = nil

Reply:
Got it